### PR TITLE
layers: Improve Consecutive Binding Updates errors

### PR DESCRIPTION
**Pull request by spencer-lunarg**
The error message before were lack luster (and found some edge case bugs too!) 

This adds better message and even added a VulkanGuide article to go along with it https://github.com/KhronosGroup/Vulkan-Guide/pull/317

### CODE REVIEW

File: layers/state_tracker/descriptor_sets.h [hunk 1]
```diff
@@ -944,10 +944,46 @@ class DescriptorSet : public StateObject, public SubStateManager<DescriptorSetSu
         DescriptorIterator &operator=(const DescriptorIterator &rhs) = default;
 
         DescriptorIterator(DescriptorSet &descriptor_set, uint32_t binding, uint32_t index = 0)
-            : iter_(descriptor_set.FindBinding(binding)), end_(descriptor_set.end()), index_(index) {}
+            : iter_(descriptor_set.FindBinding(binding)), end_(descriptor_set.end()), index_(0) {
+            if (index < (*iter_)->count) {
+                index_ = index;
+            } else {
+                // This is a consecutive binding updates and need to find first used binding
```
Comment (spencer-lunarg) [inline layers/state_tracker/descriptor_sets.h:951]: this is ugly, but this is the "rare" case (which we never tested and now do)

```
{
    binding = 0
    descriptorCount = 2
},
{
    binding = 1
    descriptorCount = 2
}

descriptor_write.dstBinding = 0;  // start a 0, but never update one in it
descriptor_write.dstArrayElement = 2; // jumps to binding 1
```

this case we need to manually update it here

File: layers/core_checks/cc_descriptor.cpp [hunk 2]
```diff
@@ -863,78 +864,61 @@ bool CoreChecks::ValidateDrawState(const vvl::DescriptorSet &descriptor_set, uin
...
-                error_str << " (" << string_VkDescriptorBindingFlags(current_binding->binding_flags)
-                          << " != " << string_VkDescriptorBindingFlags(orig_binding.binding_flags) << ")";
-            }
-        }
-
-        error_str << " so this update is invalid";
-        const char *vuid = is_copy ? "VUID-VkCopyDescriptorSet-srcSet-00349" : "VUID-VkWriteDescriptorSet-dstArrayElement-00321";
-        skip |= LogError(vuid, set.Handle(), set_loc, "%s", error_str.str().c_str());
+            auto last_binding = --current_iter;
+            return LogError(vuid, set.Handle(), binding_loc,
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_descriptor.cpp:885]: > vkUpdateDescriptorSets(): pDescriptorWrites[0].dstBinding (0) with dstArrayElement (3) went pass binding 1 (which has a descriptorCount of 1) which was the last binding in the descriptor set.

File: layers/core_checks/cc_descriptor.cpp [hunk 3]
```diff
@@ -863,78 +864,61 @@ bool CoreChecks::ValidateDrawState(const vvl::DescriptorSet &descriptor_set, uin
...
+                            binding, orig_binding.count, String(array_element_name), array_element, (*current_iter)->binding);
+        }
+    }
+
+    // Now that the arrayElement was applied, make sure each consecutive binding can handle the rest of descriptorCount
+    uint32_t updated_count = 0;
+    while (updated_count < descriptor_count) {
+        if (current_iter == set.end()) {
+            auto last_binding = --current_iter;
+            return LogError(vuid, set.Handle(), binding_loc,
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_descriptor.cpp:903]: > vkUpdateDescriptorSets(): pDescriptorWrites[0].dstBinding (1) starting at dstArrayElement (2) with descriptorCount 3 only got to updating 1 but then went pass binding 1 (which has a descriptorCount of 2) which was the last binding in the descriptor set.

File: layers/core_checks/cc_descriptor.cpp [hunk 4]
```diff
@@ -1769,35 +1758,58 @@ bool CoreChecks::ValidateWriteUpdate(const vvl::DescriptorSet &dst_set, const Vk
...
+                                     "(%" PRIu32
+                                     ") was created with %s\n"
+                                     "The descriptorCount was %" PRIu32 " and the offset of %" PRIu32 " references binding %" PRIu32
+                                     " which was created with %s.\n"
+                                     "When doing consecutive binding updates, the descriptorType must match for each binding.%s",
+                                     update.dstBinding, string_VkDescriptorType(descriptor_type), update.descriptorCount, i,
+                                     current_binding->binding, string_VkDescriptorType(current_binding->type), extra.str().c_str());
+                } else if (current_binding->stage_flags != stage_flags) {
+                    const LogObjectList objlist(update.dstSet, dst_layout->Handle());
+                    skip |= LogError("VUID-VkWriteDescriptorSet-descriptorCount-00317", objlist, write_loc.dot(Field::dstBinding),
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_descriptor.cpp:1782]: > vkUpdateDescriptorSets(): pDescriptorWrites[0].dstBinding (2) was created with VK_SHADER_STAGE_FRAGMENT_BIT
The descriptorCount was 4 and the offset of 3 references binding 4 which was created with VK_SHADER_STAGE_VERTEX_BIT.
When doing consecutive binding updates, the stageFlags must match for each binding.

File: layers/core_checks/cc_descriptor.cpp [hunk 5]
```diff
@@ -1769,35 +1758,58 @@ bool CoreChecks::ValidateWriteUpdate(const vvl::DescriptorSet &dst_set, const Vk
...
+                                     update.dstBinding, string_VkShaderStageFlags(stage_flags).c_str(), update.descriptorCount, i,
+                                     current_binding->binding, string_VkShaderStageFlags(current_binding->stage_flags).c_str());
+                } else if (current_binding->has_immutable_samplers != immutable_samplers) {
                     const LogObjectList objlist(update.dstSet, dst_layout->Handle());
-                    skip |= LogError("VUID-VkWriteDescriptorSet-descriptorCount-00318", objlist, write_loc,
-                                     "binding #%" PRIu32 " (started on dstBinding [%" PRIu32 "] + %" PRIu32
-                                     " descriptors offset) %s Immutable Samplers, which is different from the previous binding.",
-                                     current_binding->binding, update.dstBinding, i,
-                                     current_binding->has_immutable_samplers ? "has" : "doesn't have");
+                    skip |= LogError("VUID-VkWriteDescriptorSet-descriptorCount-00318", objlist, write_loc.dot(Field::dstBinding),
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_descriptor.cpp:1792]: > vkUpdateDescriptorSets(): pDescriptorWrites[0].dstBinding (0) was created with a null pImmutableSamplers
The descriptorCount was 2 and the offset of 1 references binding 1 which was created with a non-null pImmutableSamplers.
When doing consecutive binding updates, the pImmutableSamplers must all be null or non-null for each binding.

File: layers/state_tracker/descriptor_sets.h (hunk 1 above)
Comment (artem-lunarg) [reply layers/state_tracker/descriptor_sets.h:951]: I mean you probably will think it's bug in the code when see this!

Comment (spencer-lunarg) [reply layers/state_tracker/descriptor_sets.h:951]: > so the spec allows to index into the next binding? 

only if you have the same descriptor type and stage... honestly this was never working, so high chance people are not even doing this

File: layers/core_checks/cc_descriptor.cpp [hunk 6]
```diff
@@ -863,78 +864,61 @@ bool CoreChecks::ValidateDrawState(const vvl::DescriptorSet &descriptor_set, uin
...
-        error_str << " binding #" << orig_binding.binding << " with #" << update_count
-                  << " descriptors being updated but this update oversteps the bounds of this binding and the next binding is "
-                     "not consistent with current binding";
+    auto &orig_binding = **current_iter;  // save for error message
+    const char *vuid = is_copy ? "VUID-VkCopyDescriptorSet-srcSet-00349" : "VUID-VkWriteDescriptorSet-dstArrayElement-00321";
+
+    // check if srcArrayElement/dstArrayElement is so large it is skipping the the srcBinding/dstBinding
+    // If it, find the first binding being updated
+    uint32_t offset = array_element;
+    while (offset > (*current_iter)->count) {
```
Comment (artem-lunarg) [inline layers/core_checks/cc_descriptor.cpp:880]: `>=`? I was just thinking about case when array_element is equal to count (count-1 is the last index), so this should wrap to index 0 of the next binding?

Comment (spencer-lunarg) [reply layers/core_checks/cc_descriptor.cpp:880]:
```
{
    binding = 0
    descriptorCount = 2
},
{
    binding = 1
    descriptorCount = 2
}

descriptor_write.dstBinding = 0;  // start a 0, but never update one in it
descriptor_write.dstArrayElement = 2; // jumps to binding 1
```

using this example, you are right, `offset` is `2` and should be in this loop, I think I am just missing this edge case (my test had `offset` at `3` and `descriptorCount` at `1`

File: layers/state_tracker/descriptor_sets.h (hunk 1 above)
Comment (spencer-lunarg) [reply layers/state_tracker/descriptor_sets.h:951]: added more direct comment

File: layers/core_checks/cc_descriptor.cpp (hunk 6 above)
Comment (spencer-lunarg) [reply layers/core_checks/cc_descriptor.cpp]: fixed

**Comment by spencer-lunarg** [comment]
> CI Vulkan-ValidationLayers build # 20207 failed.

unrelated error
